Add a `signature` table on Darwin

andrew-d opened the pull request:
This table allows verifying the signature of files (or bundles) on
Darwin.  It also provides the signing identifier that is a part of the
signature.

I've made a couple of possibly-controversial changes here, most important of which is adding a dependency on the Cocoa framework in order to get access to NSAppKitVersionNumber.  If this is something you'd rather not support, let me know and I'd be happy to strip that portion out.

osqueryer commented:
Can one of the osquery admins reply with "ok to test" to kick off a build...

theopolis commented:
This is awesome, you rock! I'm not worried about adding another framework as long as it builds on 10.9. And the day we add something that doesn't build on 10.9 we can talk about how to go forward, hopefully that day is far out and the answer becomes more obvious. 

I am curious about the name of the table. Since this *should* work on Mach-O binaries, app bundles, and installer/pkgs should it be called `signature`? App bundles are used for a ton of things, including plugin bundles and kernel extensions, it's really neat to have 1 table to rule all of signature retrievals. A platform-agnostic signature table will also work well for Linux when reporting on signed kernel modules.

theopolis commented:
ok to test

osqueryer commented:
Build (5d91076148758d306d08b2cbacc3a25ff8a26c3f) is good! :+1:

andrew-d commented:
@theopolis Sure, I could rename it to `signature` - I found the name somewhere in the various issues requesting it, and just went with it :smile: 

The main reason I didn't make a generic table for Darwin and Linux is since there may be additional, platform-specific information we want to expose?  I'm not sure how you'd handle that (different columns on different platforms?  NULL in columns that don't apply?).

Happy to do that if it's something you want, though!

osqueryer commented:
Build (5d91076148758d306d08b2cbacc3a25ff8a26c3f) is good! :+1:

theopolis commented:
Yeah! If there is platform-specific data we can ignore those columns elsewhere. We can eventually add nullable/non-nullable hints to the column schema but for now, blank columns are fine.

I'd recommend just renaming the table (and filename for the spec), but leaving the spec within the `/specs/darwin/` folder. Then when a Linux implementation comes along we can move that spec out of the Darwin namespace.

andrew-d commented:
Okay, done!  Quick note - I haven't run the build on OS X yet (home PC is Linux), so I'll do that tomorrow and, if everything's good, let you know :8ball:

osqueryer commented:
Build (5d91076148758d306d08b2cbacc3a25ff8a26c3f) is good! :+1:

theopolis commented on osquery/tables/system/darwin/signature.mm:
```diff
@@ -0,0 +1,126 @@
+    // Note: we are explicitly *not* using is_regular_file here, since you can
+    // pass a directory path to the verification functions (e.g. for app
+    // bundles, etc.)
+    if (!boost::filesystem::exists(path)) {
```
Instead of using boost's `::exists` directly, prefer passing the `const std::string` to [`osquery::pathExists`](https://github.com/facebook/osquery/blob/master/osquery/filesystem/filesystem.cpp#L178) as you'll get exception handling for free.

theopolis commented on osquery/tables/system/darwin/signature.mm:
```diff
@@ -0,0 +1,126 @@
+  result = SecStaticCodeCheckValidityWithErrors(staticCode, csFlags, NULL, NULL);
+  if (result != 0) {
+    // If this errors, then we either don't have a signature, or it's malformed.
+    VLOG(2) << "Static code validity check failed for file: " << path;
```
Use `VLOG(1)` for everything in tables, the binaries take verbose level control away from the user and turn it into a default=WARNING/ERROR, verbose=VLOG(0-3)/INFO/WARNING/ERROR. There's also a seldom-used `TLOG`, that does not take an argument, which was designed to be the verbose logger within table implementations. :P

theopolis commented on osquery/tables/system/darwin/signature.mm:
```diff
@@ -0,0 +1,126 @@
+
+  VLOG(3) << "Everything succeeded for: " << path;
+
+done:
```
We should avoid `goto`s and the labels. It might make for a bit of duplication in release checking but that's fine. If we hit a breaking point we can switch to constructing unique pointers with custom destructors.

andrew-d replied:
I'll pick this up again tomorrow, but: if I were to create a custom pointer that simply called `CFRelease` on something, where would you like me to put it?  `osquery/core/darwin/release_ptr.h`, maybe?

andrew-d replied:
Though I guess I could just do it inline using `std::unique_ptr`... either way, I'll poke at it tomorrow :smile:

theopolis replied:
Having a facility like `shared_release_ptr` for the CF copy/create APIs would be great! If you do have time to write something like this and *have some more time*, double check the correctness with `make sanitize`. Then select from the table implementation. Otherwise, some repeated `CFReleases` are ok, we can always codemode to refactor things to `release_ptr`-style later.

theopolis commented on osquery/tables/system/darwin/signature.mm:
```diff
@@ -0,0 +1,126 @@
+                         QueryData& results) {
+  Row r;
+  OSStatus result = 0;
+  CFStringRef ident = NULL;
```
You could delay the instantiation of these pointers/const pointers until later in the method. When you do initialize them, prefer to use `nullptr`, it will make worlds of different when we run our static analysis checks!

theopolis commented:
Overall this is looking great! `make analyze` has no complaints and I don't expect any from the Memory Sanitizer either.

I was thinking, there is a bit of an implicit expectation that the executable is going to chew through CPU if one is checking signatures for application bundles. A `JOIN` from `kernel_extensions` `USING` the path would be expensive. It might be worth noting in the table spec description that signature checking is often an expensive operation. What do you think? :dancer:

andrew-d commented:
@theopolis Okay, I think I addressed all your review comments.  I've also added more tests for signature verification - specifically, ensuring that both an unsigned file and a file that is signed but has been modified are treated properly.

Let me know if there's anything else you want addressed :smile:

osqueryer commented:
Build (5d91076148758d306d08b2cbacc3a25ff8a26c3f) is good! :+1:

theopolis commented:
Perfect! One last request, can you `git rebase -i HEAD~6`, then [squash the previous 5 commits](https://ariejan.net/2011/07/05/git-squash-your-latests-commits-into-one/), so you'll end up with 1 commit for this pull request? You can `git push -f` to update the PR. We try to capture each feature add as a single commit for organization and fast reverting.

andrew-d commented:
@theopolis Done!

osqueryer commented:
Build (5d91076148758d306d08b2cbacc3a25ff8a26c3f) is good! :+1:

theopolis commented:
@andrew-d thanks again!

Aside: I might add to the testing harnesses for this later tonight. I like your approach but something bugs be about using the actual `/bin/ls` on the system that's performing the tests. I don't mind moving an `ld` or `ls` into our `./tools/tests/` directory. :P

andrew-d commented:
Sure, that would be good!  I wasn't sure what the legality was about distributing binaries from OS X, or I would have taken that route myself :books:

theopolis commented:
That's a great point! I have no clue, but we can do one better and sign a small bin ourselves and distribute that alongside. ;)
